### ISSUE TITLE
usb configuration error

### ISSUE THREAD

**Issue by ycean** [hcrest/libfreespace#8]
I am facing the problem while configuring the device.
below are the error message shown after i run " sudo mount -t usbfs /dev/bus/usb /proc/bus/usb"

**mount: mount point /proc/bus/usb does not exist**

Is there any solution for this?

**Comment by kay-hc** [comment]
Hi,

Can you edit or add /etc/udev/rules.d/99-freespace-usb.rules with the following?

# This file grants read/write ability to users in the "plugdev" group for all Hillcrest Devices  
 
# For all Hillcrest devices on USB 
SUBSYSTEM=="usb",       ATTR{idVendor}=="1d5a",     MODE="0660",    GROUP="plugdev" 
# For all Hillcrest devices on hidraw: 
KERNELS=="*:1D5A:*",     SUBSYSTEM=="hidraw",       MODE="0660",    GROUP="plugdev"

**Comment by ycean** [comment]
Hi, 
Thanks for the reply. I did tried to add all those as you mentioned, still the same error is coming out.
Is there any configuration needed for libusb?  Actually, I am confusing for the compiling step shown in the README file, which I had no idea on this step. Can lead me the step clearly?
Below is the error I had:
![screenshot from 2019-01-23 23-30-28](https://user-images.githubusercontent.com/40931878/51613497-e922ad80-1f66-11e9-8385-2b33537cd9c2.png)

Sorry, if I am asking the non-related question.

Regards,
Yeoh.

**Comment by kay-hc** [comment]
Hi,

Can you check your Python version? libfreespace requires Python v2.7.

**Comment by ycean** [comment]
I am using python2.7 , I did install python3.6 due to the installation for the doxygen is needed for it.

**Comment by kay-hc** [comment]
Hi,

Can you execute "python --version" in your console?

You can directly execute the python script from your build environment if you can get the source code generated under gen_src directory. You should be able to find freespace_codecs.c and freespace_printers.c

libfreespace/common$ python messageCodeGenerator.py -I ../include/ -s ../gen_src/ setupMessages.py

**Comment by ycean** [comment]
Hi,
The current python version is 3.7.1. 
I did try to uninstall python 3.7, yet unsuccessful . 
And I try to execute the python script, the error is coming like below:
![screenshot from 2019-01-24 16-12-31](https://user-images.githubusercontent.com/40931878/51661134-43695000-1ff3-11e9-8aa3-3f2720dfb63e.png)

I couldn't found both the source code for freespace_codecs.c and freespace_printers.c.
Sorry if I had any misunderstanding on your instruction, point me out if I did.
Your kindness would be appreciated~
Thanks.

Regards,
Yeoh

**Comment by limkx-idcc** [comment]
Hi,

Can you check your python version ("python --version") before you run the script?

**Comment by ycean** [comment]
I checked is 3.7.1.

**Comment by kay-hc** [comment]
Hi,

Python 3 is incompatible with the 2.x. libfreespace requires Python 2.x. You need to be able to switch to 2.x to be able to compile libfreespace. Please follow instructions in https://linuxconfig.org/how-to-change-from-default-to-alternative-python-version-on-debian-linux
Please note you need Python 2.7, not 3.x.

**Comment by ycean** [comment]
Yeah, I was looking for the method to change the version to 2.x , I did tried to uninstall the python 3.x , end up i did the mess that my ubuntu desktops tab were gone...thanks for the link, I am gonna try with it.

**Comment by virraaj** [comment]
Also, I am not able to sudo mount -t usbfs /dev/bus/usb /proc/bus/usb
I get this error
mount: mount point /proc/bus/usb does not exist

After searching for this error, I found [https://askubuntu.com/questions/326597/enabling-usbfs-in-ubuntu](url) here, that there is already usbfs at /sys/bus/usb/drivers/ 
Therefore I think there is no need of mounting it on /proc/bus/usb/ which was a path used in older Ubuntu versions. 

What are the requirements for getting data from FSM-9 on linux?

when I do cat /proc/bus/input/devices, this is what I get
I: Bus=0003 Vendor=1d5a Product=c0e0 Version=0111
N: Name="Hillcrest Laboratories Freespace(R) Adapter F"
P: Phys=usb-0000:00:06.0-2/input0
S: Sysfs=/devices/pci0000:00/0000:00:06.0/usb1/1-2/1-2:1.0/0003:1D5A:C0E0.0004/input/input9
U: Uniq=0001000C-0000082B
H: Handlers=mouse2 event7 
B: PROP=0
B: EV=17
B: KEY=ff0000 0 0 0 0
B: REL=103
B: MSC=10

Which means the device is already mounted.

I did the changes suggested herein /etc/udev/rules.d as well 

but when I run sudo evtest /dev/input/event7
I get the following
Input driver version is 1.0.1
Input device ID: bus 0x3 vendor 0x1d5a product 0xc0e0 version 0x111
Input device name: "Hillcrest Laboratories Freespace(R) Adapter F"
Supported events:
  Event type 0 (EV_SYN)
  Event type 1 (EV_KEY)
    Event code 272 (BTN_LEFT)
    Event code 273 (BTN_RIGHT)
    Event code 274 (BTN_MIDDLE)
    Event code 275 (BTN_SIDE)
    Event code 276 (BTN_EXTRA)
    Event code 277 (BTN_FORWARD)
    Event code 278 (BTN_BACK)
    Event code 279 (BTN_TASK)
  Event type 2 (EV_REL)
    Event code 0 (REL_X)
    Event code 1 (REL_Y)
    Event code 8 (REL_WHEEL)
  Event type 4 (EV_MSC)
    Event code 4 (MSC_SCAN)
Properties:
Testing ... (interrupt to exit)




Whereas over here I think the device is not opened, therefore we are not getting the data. Do I need any more files for installation?
I guess I am missing some installation step. Can you please suggest me

**Comment by kay-hc** [comment]
Hi,

You need to send commands to enable sensor report. Please refer to the example code in https://github.com/hcrest/libfreespace-examples
Follow the instruction in README to build.
FSM-9 uses our proprietary protocol (HCOMM) and you can find more information in the reference manual available in https://www.hillcrestlabs.com/downloads/hcomm-reference-manual

**Comment by virraaj** [comment]
> Hi,
> 
> You need to send commands to enable sensor report. Please refer to the example code in https://github.com/hcrest/libfreespace-examples
> Follow the instruction in README to build.
> FSM-9 uses our proprietary protocol (HCOMM) and you can find more information in the reference manual available in https://www.hillcrestlabs.com/downloads/hcomm-reference-manual

Hello, Thank you for your reply. 
Yes I already did those build steps but when I run any program eg. enable_mouse, or motion_example, I just get the error as Found the Device. Cannot open device. -3
And, regarding HCOMM, do I need to install some tools for this protocol?

**Comment by kay-hc** [comment]
Hi,

If you are getting the following messages, please follow the instruction below.
./motion_example: Using libfreespace 0.7.1
Scanning for Freespace devices...
Found a device. Trying to open it...
Error opening device: -3

1. Create a file for example 99-freespace-usb.rules
SUBSYSTEM=="usb", ATTR{idVendor}=="1d5a", MODE="0660", GROUP="plugdev"
KERNELS==":1D5A:", SUBSYSTEM=="hidraw", MODE="0660", GROUP="plugdev"

2. Execute the following commands
$sudo install -cv 99-freespace-usb.rules /etc/udev/rules.d/
$sudo udevadm trigger

Now you should be able to run example code ./motion_example

There is no tools you need to install for HCOMM. This is a communication protocol for our freespace devices. You will need to refer this document if you want to modify the example code.

**Comment by virraaj** [comment]
Thanks a lot. Now I can get data from ./motion_example

**Comment by kgrosero** [comment]
Hi! I'm trying to use the FSM9 device into a head tracking application.

The previous answers were very helpful for me cause I had the same problems, but I'm using the operative system MANJARO from Arch Linux. 
I create the file example 99-freespace-usb.rules but I couldnt find an equivalent command $sudo install -cv 99-freespace-usb.rules /etc/udev/rules.d/ for Arch Linux and sometimes the Error opening device: -6 happens again. 

Could you please suggest me a way to install the .rules file in Linux?
